src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -4433,56 +4401,112 @@ CreateMapQueryString(MapMergeJob *mapMergeJob, Task *filterTask,
 {
 	uint64 jobId = filterTask->jobId;
 	uint32 taskId = filterTask->taskId;
+	char *resultNamePrefix = PartitionResultNamePrefix(jobId, taskId);
 
 	/* wrap repartition query string around filter query string */
 	StringInfo mapQueryString = makeStringInfo();
 	char *filterQueryString = TaskQueryString(filterTask);
-	char *filterQueryEscapedText = quote_literal_cstr(filterQueryString);
 	PartitionType partitionType = mapMergeJob->partitionType;
 
 	Var *partitionColumn = mapMergeJob->partitionColumn;
 	Oid partitionColumnType = partitionColumn->vartype;
-	char *partitionColumnTypeFullName = format_type_be_qualified(partitionColumnType);
-	int32 partitionColumnTypeMod = partitionColumn->vartypmod;
 
 	ShardInterval **intervalArray = mapMergeJob->sortedShardIntervalArray;
 	uint32 intervalCount = mapMergeJob->partitionCount;
 
 	if (partitionType == DUAL_HASH_PARTITION_TYPE)
 	{
 		partitionColumnType = INT4OID;
-		partitionColumnTypeMod = get_typmodin(INT4OID);
 		intervalArray = GenerateSyntheticShardIntervalArray(intervalCount);
 	}
 	else if (partitionType == SINGLE_HASH_PARTITION_TYPE)
 	{
 		partitionColumnType = INT4OID;
-		partitionColumnTypeMod = get_typmodin(INT4OID);
 	}
 
-	ArrayType *splitPointObject = SplitPointObject(intervalArray, intervalCount);
-	StringInfo splitPointString = ArrayObjectToString(splitPointObject,
-													  partitionColumnType,
-													  partitionColumnTypeMod);
+	Oid intervalTypeOutFunc = InvalidOid;
+	bool intervalTypeVarlena = false;
+	ArrayType *minValueArray = NULL;
+	ArrayType *maxValueArray = NULL;
 
-	char *partitionCommand = NULL;
-	if (partitionType == RANGE_PARTITION_TYPE)
-	{
-		partitionCommand = RANGE_PARTITION_COMMAND;
-	}
-	else
-	{
-		partitionCommand = HASH_PARTITION_COMMAND;
-	}
+	getTypeOutputInfo(partitionColumnType, &intervalTypeOutFunc, &intervalTypeVarlena);
+
+	ShardMinMaxValueArrays(intervalArray, intervalCount, intervalTypeOutFunc,
+						   &minValueArray, &maxValueArray);
+
+	StringInfo minValuesString = ArrayObjectToString(minValueArray, TEXTOID,
+													 InvalidOid);
+	StringInfo maxValuesString = ArrayObjectToString(maxValueArray, TEXTOID,
+													 InvalidOid);
+
+	char *partitionMethodString = partitionType == RANGE_PARTITION_TYPE ?
+								  "range" : "hash";
+
+	/* TODO: determine binary safety automatically */
+	bool useBinaryFormat = BinaryWorkerCopyFormat;
+
+	/*
+	 * NULL values never match a join condition, so it's ok to skip them.
+	 */
+	bool skipNullPartitionColumnValue = true;
```

Comment:
I guess this assumes we only support INNER joins? Because with left joins NULL values can be on the output? Or, am I missing something here?

```SQL
create table test (a int);
insert into test VALUES (NULL);
select * from test t1 LEFT JOIN test t2 USING (a);
 a 
---
  
(1 row)

```

Reply:
Yes, but the output of worker_partition_query_result is the input to the join, not the output.

The result is the same if you do  `select * from test t1 LEFT JOIN (select * from test limit 0) t2 USING (a);` . The NULL input never matches the join condition, so it might as well not be there.

Reply:
Actually, you're right, we should preserve NULL values on the outside, but can skip them on the inside. I changed to writing NULL values to the first partition and omitted the skipping logic for now.